src/platform/datapath_winuser.c

```diff
@@ -4084,12 +4114,20 @@ CxPlatDataPathProcessCqe(
         break;
     }
     case CXPLAT_CQE_TYPE_SOCKET_IO: {
-        CXPLAT_SOCKET_PROC* SocketProc =
-            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), CXPLAT_SOCKET_PROC, IoSqe);
-        if (Cqe->lpOverlapped == &SocketProc->IoSqe.Sqe) {
-            CxPlatDataPathSocketProcessReceiveCompletion(SocketProc, Cqe);
-        } else {
-            CxPlatDataPathSocketProcessSendCompletion(SocketProc, Cqe);
+        DATAPATH_IO_SQE* Sqe =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), DATAPATH_IO_SQE, DatapathSqe);
+
+        switch (Sqe->IoType) {
```

Comment:
This is the motivation of the PR: each IO needs to be able to provide an arbitrary completion context. Without this pattern, we were stuck with the fragile completion pattern before: one code path for when the overlapped IO matched the IO completion key, a separate code path for when it _didn't_ match the IO completion key. We then had an assortment of thread-the-needle stateful checks and magic parameter values to distinguish between completions.